hikari/traits.py

```diff
@@ -60,42 +57,64 @@
     from hikari.api import rest as rest_
     from hikari.api import shard as gateway_shard
     from hikari.api import voice as voice_
-    from hikari.events import base_events  # noqa F401 - Unused (False positive)
 
-EventT_co = typing.TypeVar("EventT_co", bound="base_events.Event", covariant=True)
-"""Type-hint for a covariant event implementation instance.
+    _T = typing.TypeVar("_T")
 
-This will bind to the bound event type, or any subclass of that type.
-"""
+_Protocol = NotImplemented
 
-EventT_inv = typing.TypeVar("EventT_inv", bound="base_events.Event")
-"""Type-hint for an invariant event implementation instance.
 
-This will bind to the bound event type only. Subclasses and superclasses will
-not be matched.
-"""
+class _FastProtocol(abc.ABCMeta):
+    @staticmethod
+    def __new__(
+        mcs: typing.Type[_T],
+        cls_name: str,
+        bases: typing.Tuple[typing.Type[typing.Any], ...],
+        namespace: typing.Dict[str, typing.Any],
+    ) -> _T:
+        global _Protocol
 
-PredicateT = typing.Callable[[EventT_co], typing.Union[bool, typing.Coroutine[typing.Any, typing.Any, bool]]]
-"""Type-hint for an event waiter predicate.
+        if _Protocol is NotImplemented:
+            if cls_name != "Protocol":
+                raise TypeError("First instance of _FastProtocol must be Protocol")
 
-This should be a function or coroutine function that consumes a covariant
-instance of the bound event type and returns a boolean that matches
-`builtins.True` if the event is a match for the waiter, or `builtins.False`
-otherwise.
-"""
+            namespace["_is_protocol"] = True
+            namespace["_attributes_"] = ()
+            return (_Protocol := super().__new__(mcs, cls_name, bases, namespace))
 
-# Fixme: shouldn't this be using covariance instead of invariance?
-AsyncCallbackT = typing.Callable[[EventT_inv], typing.Coroutine[typing.Any, typing.Any, None]]
-"""Type-hint for an asynchronous coroutine function event callback.
+        if _Protocol not in bases:
+            namespace["_is_protocol"] = False
 
-This should consume a single argument: the event that was dispatched.
+        else:
+            attributes = tuple(name for name in namespace if not name.startswith("_"))
 
-This is not expected to return anything.
-"""
+            for base in bases:
+                if not issubclass(base, _Protocol):
+                    raise TypeError(f"Protocols can only inherit from other protocols, got {base!r}")
+                attributes += base._attributes_
 
+            namespace["_is_protocol"] = True
+            namespace["_attributes_"] = attributes
 
-@typing.runtime_checkable
-class NetworkSettingsAware(typing.Protocol):
+        return super().__new__(mcs, cls_name, bases, namespace)
+
+    def __instancecheck__(cls, obj: typing.Any) -> bool:
+        if not cls._is_protocol:
+            return super().__instancecheck__(obj)
+
+        for i in cls._attributes_:
+            if not hasattr(obj, i):
```

Comment:
how about a `return all(...)` statement here?

Reply:
There is a reason I didn't go for that, and that is because this way we can quickly fail as soon as we see that one of the attributes is not there.

Reply:
`all` returns on the first item in the iterator/iterable which isn't truthful (and `any` returns on the first which is truthful)
```py
>>> def divisible_by_4(i) -> bool:
...     print(i)
...     return i % 4 == 0
... print(all(divisible_by_4(i) for i in (4,8,12, 16, 17, 20, 24, 28)))
4
8
12
16
17
False
```

Reply:
Seems to be slower, so will keep what I got rn :+1: